src/helpers/mock.rs

```diff
@@ -0,0 +1,286 @@
+        })?;
+
+        Ok(())
+    }
+
+    async fn receive<T: Message>(
+        &mut self,
+        source: Identity,
+        record: RecordId,
+    ) -> Result<T, Error> {
+        let envelope = self.controller.receive(source, self.step, record).await;
+        let obj: T = serde_json::from_slice(&envelope.payload).unwrap();
+
+        Ok(obj)
+    }
+
+    fn identity(&self) -> Identity {
+        self.controller.identity
+    }
+}
+
+impl<S> Clone for Controller<S> {
+    fn clone(&self) -> Self {
+        Self {
+            identity: self.identity,
+            peers: self.peers.clone(),
+            connections: Arc::clone(&self.connections),
+            buf: Arc::clone(&self.buf),
+        }
+    }
+}
+
+impl<S: Step> Controller<S> {
+    fn launch(
+        identity: Identity,
+        control_tx: HashMap<Identity, Sender<ControlMessage<S>>>,
+        control_rx: Receiver<ControlMessage<S>>,
+    ) -> Self {
+        let controller = Self {
+            identity,
+            connections: Arc::new(Mutex::new(HashMap::new())),
+            buf: Arc::new(Mutex::new(HashMap::new())),
+            peers: control_tx,
+        };
+
+        controller.start(control_rx);
+
+        controller
+    }
+
+    fn start(&self, mut rx: Receiver<ControlMessage<S>>) {
+        tokio::spawn({
+            let controller = self.clone();
+            async move {
+                while let Some(msg) = rx.recv().await {
+                    match msg {
+                        ControlMessage::ConnectionRequest(peer, step, peer_connection) => {
+                            controller.connect(peer, step, peer_connection);
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    fn connect(&self, peer: Identity, step: S, mut rx: Receiver<MessageEnvelope>) {
+        assert_ne!(self.identity, peer);
+
+        // start listening for incoming messages and move them from channel to the buffer
+        tokio::spawn({
+            let buf = Arc::clone(&self.buf);
+            async move {
+                while let Some(msg) = rx.recv().await {
+                    let mut buf = buf.lock().unwrap();
+                    buf.entry((peer, step)).or_insert_with(Vec::new).push(msg);
+                }
+            }
+        });
+    }
+
+    async fn get_connection(&self, peer: Identity, step: S) -> Sender<MessageEnvelope> {
+        assert_ne!(self.identity, peer);
+
+        loop {
+            // Depending on connection status, request a new connection, spin and wait for
+            // connection acknowledgment from peer or return the sender end of connection
+            // if it is ready
+            let control_message = {
+                let mut connections = self.connections.lock().unwrap();
+                let conn_state = connections
+                    .entry((peer, step))
+                    .or_insert_with(|| ConnectionState::Listen);
```

Comment:
```suggestion
                    .or_insert(ConnectionState::Listen);
```

Reply:
Does anybody know why this suggestion is not visible from the files tab??

Reply:
if you've made a change to that line, it'll only show up on commits before the change. so you'd have to pick an earlier commit in the files tab to see it